### ISSUE TITLE
[ Regression ] HHVM 4.58.0 does not allow for tuples to appear in attribute expressions

### ISSUE THREAD

**Issue by lexidor** [facebook/hhvm#8704]
**Describe the bug**

7ce3a3b added a typechecker check to prevent attribute expressions. This made `tuple()` expressions illegal in attribute context, where they were valid before.

This is now also a runtime error, because attributes are processed in `hh_single_compile`, instead of at runtime.

**Standalone code, or other way to reproduce the problem**

```HACK
final class TakesTuple implements \HH\FunctionAttribute {
    public function __construct((int) $_) {}
}

final class TakesShape implements \HH\FunctionAttribute {
    public function __construct(shape('a' => int) $_) {}
}

<<TakesShape(shape('a' => 0))>>
function thing_shape(): void {}

<<TakesTuple(tuple(1))>>
function thing_tuple(): void {}

<<__EntryPoint>>
function main(): void {
    $thing_shape = new \ReflectionFunction('thing_shape');
    $thing_tuple = new \ReflectionFunction('thing_tuple');
    \var_dump($thing_shape->getAttributes());
    \var_dump($thing_tuple->getAttributes());
}
```

1. Copy the code into an empty hack project.
2. Run the typechecker.
3. Interpret the file with hhvm.

**Expected behavior**

`tuple()` expressions in attribute context should remain valid.

**Screenshots**

![image](https://user-images.githubusercontent.com/31805625/82330849-a0a4c400-99e3-11ea-8e4c-f363aa2d0568.png)

**Desktop (please complete the following information):**
 - OS: Ubuntu 20.04 with Docker containers
 - HHVM Version: hhvm 4.57.0 `-->` 4.58.0

```
HipHop VM 4.57.0 (rel)
Compiler: 1589216405_387415496
Repo schema: d1ae8e21bf3419a65f12a010527485564e719d07
hackc-f1be4ccd34a71c94959e53ce98357d79f9a86e04-4.57.0

HipHop VM 4.58.0 (rel)
Compiler: 1589823152_618326559
Repo schema: d1ae8e21bf3419a65f12a010527485564e719d07
hackc-fb75b3ba956958b0e72ab68c65f2a8261e31fe3a-4.58.0
```

**Additional context**

Command to spin this code up in many different hhvm versions on the same machine.
```
sudo docker run --tty --interactive -v /path/to/code:/mnt/code hhvm/hhvm:4.VERSION.0 /bin/bash -l
cd /mnt/code
hh_client
```

Commit message of 7ce3a3b
```
Summary:
Attribute arguments that are not literals pass the typechecker, but cause HHVM to fatal. We want to warn users about invalid attributes during the parsing pass so that they will not be able to ship code to production that will fail.

The word "literal" here is used pretty loosely and includes:
* Regular Literals (eg string, int, etc)
* Collections of literals (eg vec)
* Concatenated strings
* `::class` strings
* shapes

HHVM seems to support numeric arithmetic too, but I've decided not to allow it.
```

**Comment by fredemmott** [comment]
FB T67208213

**Comment by Wilfred** [comment]
588cb0e5bea9545a53a912a1d5bb3118985f830f

**Comment by lexidor** [comment]
Diff merged into hhvm version [4.59.0](https://hhvm.com/blog/2020/05/26/hhvm-4.59.html).
